src/libertem/io/dataset/base/backend_buffered.py

```diff
@@ -0,0 +1,262 @@
+                           out_decoded, native_dtype, do_zero,
+                           origin, shape, ds_shape, offsets):
+        if do_zero:
+            out_decoded[:] = 0
+        for rr_idx in range(tile_read_ranges.shape[0]):
+            rr = tile_read_ranges[rr_idx]
+            if rr[1] == rr[2] == 0:
+                break
+            buf = buffers[rr[0]]
+            offset = offsets[rr[0]]  # per-file offset of start of read buffer
+            decode(
+                inp=buf[rr[1] - offset:rr[2] - offset],
+                out=out_decoded,
+                idx=rr_idx,
+                native_dtype=native_dtype,
+                rr=rr,
+                origin=origin,
+                shape=shape,
+                ds_shape=ds_shape,
+            )
+        return out_decoded
+    return _buffered_tilereader
+
+
+@numba.njit(inline='always')
+def block_get_min_fill_factor(rrs):
+    """
+    Try to find out how sparse the given read ranges are, per file.
+
+    Returns the smallest fill factor and maximum required buffer size.
+    """
+    # FIXME: replace with fixed arrays? If this fn stands out on a profile...
+    min_per_file = {}
+    max_per_file = {}
+    payload_bytes = {}
+    for tile_idx in range(rrs.shape[0]):
+        for part_idx in range(rrs.shape[1]):
+            part_rr = rrs[tile_idx, part_idx]
+            fileno = part_rr[0]
+
+            if part_rr[2] == part_rr[1] == 0:
+                continue
+
+            if fileno not in min_per_file:
+                min_per_file[fileno] = part_rr[1]
+            else:
+                min_per_file[fileno] = min(part_rr[1], min_per_file[fileno])
+
+            if fileno not in max_per_file:
+                max_per_file[fileno] = part_rr[2]
+            else:
+                max_per_file[fileno] = max(part_rr[2], max_per_file[fileno])
+
+            if fileno not in payload_bytes:
+                payload_bytes[fileno] = 0
+            payload_bytes[fileno] += part_rr[2] - part_rr[1]
+
+    min_fill_factor = -1
+    max_buffer_size = 0
+
+    for k in min_per_file:
+        read_size = max_per_file[k] - min_per_file[k]
+        fill_factor = payload_bytes[k] / (read_size)
+        # print(k, fill_factor, max_per_file[k], min_per_file[k], payload_bytes[k])
+        if fill_factor < min_fill_factor or min_fill_factor == -1:
+            min_fill_factor = fill_factor
+        if read_size > max_buffer_size:
+            max_buffer_size = read_size
+    return min_fill_factor, max_buffer_size, min_per_file, max_per_file
+
+
+class BufferedBackend(IOBackend, id_="buffered"):
+    def __init__(self, max_buffer_size=16*1024*1024, sparse_threshold=0.9):
+        """
+        I/O backend using a buffered reading strategy. Useful for slower
+        media like HDDs, where seeks cause performance drops.
+
+        This does not perform optimally on SSDs, there
+        :class:`~libertem.io.dataset.base.MMapBackend` should be used
+        instead.
```

Comment:
Actually, buffered reading performed very well on my Windows laptop with a SATA SSD under memory pressure. As far as I can tell, the biggest drawback is slower performance for files that are in the file system cache. In practice only relevant for systems with pretty large memory?

Reply:
I've updated the wording to hopefully clarify that this backend is not worse in all situations, instead hinting that it really depends on the circumstances.